[1.18.x] Model Providers: The Long (#442)
MinecraftForge/Documentation: docs/datagen/client/modelproviders.md

Model Generation
================

[Models] can be generated for models or block states by default. Each provides a method of generating the necessary JSONs (`ModelBuilder#toJson` for models and `IGeneratedBlockState#toJson` for block states). After implementation, the [associated providers][provider] must be [added][datagen] to the `DataGenerator`.

Model Files
-----------

A `ModelFile` acts as the base for all models referenced or generated by a provider. Each model file stores the location relative to the `models` subdirectory and can assert whether the file exists.

### Existing Model Files

`ExistingModelFile` is a subclass of `ModelFile` which checks via [`ExistingFileHelper#exists`][efh] whether the model already exists within the `models` subdirectory. All non-generated models are usually referenced through `ExistingModelFile`s.

### Unchecked Model Files

`UncheckedModelFile` is a subclass of `ModelFile` which assumes the specified model exists in some location.

!!! note

>     There should be no cases where an `UncheckedModelFile` is used to reference a model. If there is, then the associated resources are not properly being tracked by `ExistingFileHelper`.

Reviewer: Should we even mention the existence of `UncheckedModelFile` in the docs?

Author: I don't see a reason not to. It's the same reason why we mention `OnlyIn`: explain what it is and that it shouldn't be used.

Model Builders
--------------

A `ModelBuilder` represents a to-be-generated `ModelFile`. It contains all the data about a model: its parent, faces, textures, transformations, lighting, and [loader].

> !!! tip
>     While a complex model can be generated, it is recommended that those models be constructed using a modeling software beforehand and generate children of the complex model with the specified textures applied through the associated provider.

Reviewer: Could be rewritten to be a bit less awkward with the faulty parallelism of "be constructed" and "generate".
```suggestion
!!! tip
    While a complex model can be generated, it is recommended that those models be constructed using a modeling software beforehand. Then, datageneration can be used to generate children models of the complex model with specific textures overriding the ones in the parent complex model.
```

> The parent (via `ModeBuilder#parent`) of the builder can be any `ModelFile`: generated or existing. Generated files are added to `ModelProvider`s as soon as the builder is created. The builder itself can be passed in as a parent, or the `ResourceLocation` can supplied alternatively.

Reviewer: Typo: `ModeBuilder#parent` -> `ModelBuilder#parent`
```suggestion
The parent (via `ModelBuilder#parent`) of the builder can be any `ModelFile`: generated or existing. Generated files are added to `ModelProvider`s as soon as the builder is created. The builder itself can be passed in as a parent, or the `ResourceLocation` can supplied alternatively.
```

!!! warning
    If the parent is not generated before the child model when passing in a `ResourceLocation`, then an exception will be thrown.

> Each element (via `ModeBuilder#element`) within a model is defined as cube using two three-dimensional points (`ElementBuilder#from` and `#to` respectively) where each axis is limited to the values `[-16,32]`. Each face (`ElementBuilder#face`) of the cube can specify when the face is culled (`FaceBuilder#cullface`), [tint index][color](`FaceBuilder#tintindex`), texture reference from the `textures` keys(`FaceBuilder#texture`), UV coordinate on the texture(`FaceBuilder#uvs`), and rotation in 90 degree intervals(`FaceBuilder#rotation`).

Reviewer: Typo: `ModeBuilder#parent` -> `ModelBuilder#parent`. Missing spaces before some parenthetical expressions.
```suggestion
Each element (via `ModelBuilder#element`) within a model is defined as cube using two three-dimensional points (`ElementBuilder#from` and `#to` respectively) where each axis is limited to the values `[-16,32]`. Each face (`ElementBuilder#face`) of the cube can specify when the face is culled (`FaceBuilder#cullface`), [tint index][color] (`FaceBuilder#tintindex`), texture reference from the `textures` keys (`FaceBuilder#texture`), UV coordinate on the texture (`FaceBuilder#uvs`), and rotation in 90 degree intervals (`FaceBuilder#rotation`).
```

Regarding the limitation of `[-16,32]`: where in the code is this sourced from? Additionally, perhaps it would be good to include prose description of the limitation, for those unfamiliar with interval notation.

Author: See either `ElementBuilder#validateCoordinate`, or more accurately, `BlockElement$Deserializer#getFrom` and `#getTo`.

> !!! note
>     It recommended for block models that any axis that needs to be outside a `[0,16]` bound should be a separate model for a multiblock structure.

Reviewer: ```suggestion
!!! note
    It recommended for block models which have elements that exceed a bound of `[0,16]` on any axis to separate into multiple blocks, such as for a multiblock structure, to avoid lighting and culling issues.
```

Each cube can additionally be rotated (`ElementBuilder#rotation`) around a specified point (`RotationBuilder#origin`) for a given axis (`RotationBuilder#axis`) in 22.5 degree intervals (`RotationBuilder#angle`). The cube can scale all faces in relation to the entire model as well (`RotationBuilder#rescale`). The cube can also determine whether its shadows should be rendered (`ElementBuilder#shade`).

!!! important

>     Elements can be rotated on different axes on 90 degree intervals. Only one axis per cube can use the 22.5 degree interval.

Reviewer: Could you please point where this is sourced from?

Author: Honestly, I can't find it anymore. I think I may conflated two different portions of the model.

Model Providers
---------------

The `ModelProvider` subclasses are responsible for generating the constructed `ModelBuilder`s. The provider takes in the generator, mod id, subdirectory in the `models` folder to generate within, a `ModelBuilder` factory, and the existing file helper. Each provider subclass must implement `#registerModels`.

The provider contains basic methods which either create the `ModelBuilder` or provides convenience for getting texture or model references:

Method               | Description
:---:                | :---
`getBuilder`         | Creates a new `ModelBuilder` within the provider's subdirectory for the given mod id.
`withExistingParent` | Creates a new `ModelBuilder` for the given parent. Should be used when the parent is not generated by the builder.
`mcLoc`              | Creates a `ResourceLocation` for the path in the `minecraft` namespace.
`modLoc`             | Creates a `ResourceLocation` for the path in the given mod id's namespace.

Additionally, there are several helpers for easily generating common models using vanilla templates. Most are for block models with only a few being universal.

!!! note

>     Although the models are within a specific subdirectory, that does **not** mean that the model cannot be referenced by a model in another subdirectory. Usually, it is indicative of that model being using for that type of object.

Reviewer: Grammar error with `being using`.
```suggestion
    Although the models are within a specific subdirectory, that does **not** mean that the model cannot be referenced by a model in another subdirectory. Usually, it is indicative of that model being used for that type of object.
```

### `BlockModelProvider`

The `BlockModelProvider` is used for generating block models via `BlockModelBuilder` in the `block` folder. Block models should typically parent `minecraft:block/block` or one of its children models for use with item models.

!!! note
    Block models and its item model counterpart are typically not generated through a direct subclass of `BlockModelProvider` and `ItemModelProvider` but through [`BlockStateProvider`][blockstateprovider].

### `ItemModelProvider`

> The `ItemModelProvider` is used for generating block models via `ItemModelBuilder` in the `item` folder. Most item models parent `item/generated` and use `layer0` to specify their texture, which can be done using `#singleTexture`

Reviewer: Should this reference the possible textures of `layer0` through `layer4`?

The provider contains basic methods for generating block state JSONs and block models. Item models must be generated separately as a block state JSON may define multiple models to use in different contexts. There are a number of common methods, however, that that the modder should be aware of when dealing with more complex tasks:

Method            | Description
:---:             | :---
`models`          | Gets the [`BlockModelProvider`][blockmodels] used to generate the item block models.
`itemModels`      | Gets the [`ItemModelProvider`][itemmodels] used to generate the item block models.
`modLoc`          | Creates a `ResourceLocation` for the path in the given mod id's namespace.
`mcLoc`           | Creates a `ResourceLocation` for the path in the `minecraft` namespace.
`blockTexture`    | References a texture within `textures/block` which has the same name as the block.
`simpleBlockItem` | Creates an item model for a block given the associated model file.

> A block state JSON is made up of variants or conditions. Each variant or condition references a `ConfiguredModelList`: a list of `ConfiguredModel`s. Each configured model contains the model file (via `ConfiguredModel$Builder#modelFile`), the x and y rotation in 90 degree intervals (via `#rotationX` and `rotationY` respectively), whether the texture can rotate when the model is rotated by the block state JSON (via `#uvLock`), and the weight of the model appearing compared to other models in the list (via `#weight`).

Reviewer: Perhaps the references to the `x` and `y` axes should be capitalized?

    .rotationY(90) // Rotates 'hModelFile' 90 degrees on the Y axis
    .addModel(); // Finalizes models when AXIS = X

// EXAMPLE_BLOCK_2: Has Property BlockStateProperties#HORIZONTAL_FACING
this.getVariantBuilder(EXAMPLE_BLOCK_2) // Get variant builder
  .forAllStates(state -> // For all possible states
    ConfiguredModel.builder() // Creates configured model builder
      .modelFile(modelFile) // Can show 'modelFile'
      .rotationY((int) state.getValue(HORIZONTAL_FACING).toYRot()) // Rotates 'modelFile' on the Y axis depending on the property
      .build() // Creates the array of configured models
  );

// EXAMPLE_BLOCK_3: Has Properties BlockStateProperties#HORIZONTAL_FACING, BlockStateProperties#WATERLOGGED
this.getVariantBuilder(EXAMPLE_BLOCK_3) // Get variant builder
  .forAllStatesExcept(state -> // For all HORIZONTAL_FACING states
    ConfiguredModel.builder() // Creates configured model builder
      .modelFile(modelFile) // Can show 'modelFile'
      .rotationY((int) state.getValue(HORIZONTAL_FACING).toYRot()) // Rotates 'modelFile' on the Y axis depending on the property
      .build() // Creates the array of configured models

>   , WATERLOGGED); // Ignores WATERLOGGED property

Reviewer: The placement of this comma looks a bit awkward to me.

`DynamicBucketModelBuilder`       | `#begin`       | Generates a bucket model for the specified fluid.
`CompositeModelBuilder`           | `#begin`       | Generates a model composed of models.
`ItemLayersModelBuilder`          | `#begin`       | Generates a Forge implementation of an `item/generated` model.
`SeparatePerspectiveModelBuilder` | `#begin`       | Generates a model which changes based on the specified [perspective].
`OBJLoaderBuilder`                | `#begin`       | Generates an [OBJ model][obj].
`MultiLayerModelBuilder`          | `#begin`       | Generates a model made up of models in different rendering layers.

```java
// For some BlockModelBuilder builder
builder.customLoader(OBJLoaderBuilder::begin) // Custom loader 'forge:obj'
  .modelLocation(new ResourceLocation(MOD_ID, "models/block/model.obj")) // Set the OBJ model location
  .flipV(true) // Flips the V coordinate in the supplied .mtl texture
  .end() // Finish custom loader configuration
.texture("particle", new ResourceLocation("block/dirt")) // Set particle texture to dirt
.texture("texture0", new ResourceLocation("block/dirt")); // Set 'texture0' texture to dirt

> ```

Reviewer: What about `modLoc` and `mcLoc`?
